Flush errors swallowed — tests pass while data integrity fails

## Summary

Internal flush failures from `compress_and_flush_data_` are logged but not propagated back to the caller of `ZarrStream_append` on the call that triggered them. They only surface on a *subsequent* `append` via the `error_` member — and if no such call occurs, they never surface at all. The test passes even though an error was logged.

## Evidence

Seen in a recent CI run on `macos-latest` during `test_append_throws_on_overflow`:

```
[ERROR] array.cpp:913 compress_and_flush_data_: Failed to flush data:
  Failed to open file: '/private/var/.../test_append_throws_on_overflow.zarr/c/0/0/0':
  No such file or directory
python/tests/test_stream.py::test_append_throws_on_overflow PASSED
```

The first `stream.append(data)` had an internal flush failure; its zarr chunk `c/0/0/0` was never written. The test still passed because `pytest.raises(RuntimeError)` only verifies the *second* (deliberately-overflowing) append throws.

## Root cause (sketch)

`src/streaming/array.cpp:910-914` runs the flush job either on the thread pool or inline. On failure it does `LOG_ERROR(err); all_successful.fetch_and(success)` and `promise->set_value()`, but the `all_successful` result only lands in `error_` asynchronously via the frame-queue thread. The append that *triggered* the flush has already returned `Success` by then.

## Proposed fix

Surface flush failures synchronously on the triggering call — wait on outstanding futures and convert to an error return before `ZarrStream_append` returns. Alternatively, at minimum, surface them at `ZarrStream_destroy`/`close` time so a `close` after a failed flush raises.

## Related

- `python/tests/test_stream.py::test_append_throws_on_overflow` — the test that surfaced this.

--- a/src/streaming/zarr.stream.cpp
+++ b/src/streaming/zarr.stream.cpp
@@ -1656,21 +1656,49 @@ ZarrStream_s::process_frame_queue_()
             it == output_arrays_.end()) {
             // If we have gotten here, something has gone seriously wrong
             set_error_("Output node not found for key: '" + output_key + "'");
-            std::unique_lock lock(frame_queue_mutex_);
-            frame_queue_finished_cv_.notify_all();
+            {
+                std::unique_lock lock(frame_queue_mutex_);
+                process_frames_ = false;
+                frame_queue_->clear();
+                frame_queue_not_full_cv_.notify_all();
+                frame_queue_empty_cv_.notify_all();
+                frame_queue_finished_cv_.notify_all();
+            }
             return;
         } else {
             auto& output_node = it->second;
 
             size_t n_bytes;
-            if (const auto result =
+            bool write_ok = false;
+            std::string err_msg;
+            try {
+                const auto result =
                   output_node.array->write_frame(frame, n_bytes);
-                result != zarr::WriteResult::Ok) {
-                // TODO (aliddell): retry on WriteResult::PartialWrite
-                set_error_("Failed to write frame to writer for key: " +
-                           output_key);
-                std::unique_lock lock(frame_queue_mutex_);
-                frame_queue_finished_cv_.notify_all();
+                if (result != zarr::WriteResult::Ok) {
+                    // TODO (aliddell): retry on WriteResult::PartialWrite
+                    err_msg = "Failed to write frame to writer for key: " +
+                              output_key;
+                } else {
+                    write_ok = true;
+                }
+            } catch (const std::exception& exc) {
+                // A failed compress/flush inside write_frame throws via
+                // CHECK(); surface it back to the user so the next append()
+                // sees the error instead of the failure being swallowed.
+                err_msg = "Exception while writing frame for key '" +
+                          output_key + "': " + exc.what();
+            }
+
+            if (!write_ok) {
+                set_error_(err_msg);
+                {
+                    std::unique_lock lock(frame_queue_mutex_);
+                    process_frames_ = false;
+                    frame_queue_->clear();
+                    frame_queue_not_full_cv_.notify_all();
+                    frame_queue_empty_cv_.notify_all();
+                    frame_queue_finished_cv_.notify_all();
+                }
                 return;
             }
         }
@@ -1710,11 +1738,18 @@ ZarrStream_s::finalize_frame_queue_()
         frame_queue_not_full_cv_.notify_all();
     }
 
-    // Wait for frame processing to complete
+    // Wait for frame processing to complete. Use a timeout so the destructor
+    // cannot hang indefinitely if a future change reintroduces a missing
+    // notify on one of the worker-thread early-exit paths.
     std::unique_lock lock(frame_queue_mutex_);
-    frame_queue_finished_cv_.wait(lock, [this] {
-        return frame_queue_processing_done_.load() || frame_queue_->empty();
-    });
+    if (!frame_queue_finished_cv_.wait_for(
+          lock, std::chrono::seconds(30), [this] {
+              return frame_queue_processing_done_.load() ||
+                     frame_queue_->empty();
+          })) {
+        LOG_ERROR("Timed out waiting for frame queue to finalize after 30s; "
+                  "proceeding with shutdown");
+    }
 }
 
 bool

--- a/src/streaming/zarr.stream.cpp
+++ b/src/streaming/zarr.stream.cpp
@@ -1743,7 +1743,7 @@ ZarrStream_s::finalize_frame_queue_()
     // notify on one of the worker-thread early-exit paths.
     std::unique_lock lock(frame_queue_mutex_);
     if (!frame_queue_finished_cv_.wait_for(
-          lock, std::chrono::seconds(30), [this] {
+          lock, std::chrono::seconds(300), [this] {
               return frame_queue_processing_done_.load() ||
                      frame_queue_->empty();
           })) {

